lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1487,10 +1487,15 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
     /// use or capture of "self." for qualification of member references.
     static bool isClosureRequiringSelfQualification(
                   const AbstractClosureExpr *CE) {
+      auto AFR = AnyFunctionRef(const_cast<AbstractClosureExpr *>(CE));
+      // If we don't have a type for this closure yet, don't offer diagnostics
+      // for required 'self'. We will have another chance when (if) the closure
+      // successfully typechecks.
+      if (!AFR.hasType())
+        return false;
       // If the closure's type was inferred to be noescape, then it doesn't
       // need qualification.
-      return !AnyFunctionRef(const_cast<AbstractClosureExpr *>(CE))
-               .isKnownNoEscape();
+      return !AFR.isKnownNoEscape();
```

Comment:
I don't think it's really necessary to `const_cast` and wrap in `AnyFunctionRef` to access this info, if you're already doing the `hasType()` or `getType()->hasError()` checks here. I think you can just do:

```c++
auto closureTy = CE->getType();
if (!closureTy || closureTy->hasError())
  return false;
return !closureTy->castTo<AnyFunctionType>()->isNoEscape();
```

Reply:
No I think that makes sense. I think you're right that `isKnownNoEscape` is just the wrong thing to check here, since what we really want is "`isKnownPotentiallyEscaping`". Since `isKnownNoEscape` is used elsewhere I think your suggested adjustment is a good way to go.